Decaffeinate Tests

curran opened the pull request:
Closes #7

curran commented:
At this point, this is a rough first pass at the change, presenting the raw output from the automated tool https://github.com/decaffeinate/decaffeinate .

Looking through the resulting code, the added verbosity may benefit from a [Prettier](https://prettier.io/) pass (e.g. some lines are way too long now).

curran commented:
I ran `prettier --single-quote ./*.js --write` in the `tests` directory, which resulted in a huge improvement.

curran commented:
The remaing task here is to do a manual pass over all the generated JS files and clean them up. The generated `decaffeinate suggestions` comments may be helpful with this.

I'm willing to do this last leg of work as well, but I'd like to hear from you @josephg on the general approach of this PR. What do you think? Would this be a welcome change? Thanks!

josephg commented:
Yep this looks great! I do appreciate this change, and like you I'm curious why the tests are failing. I'd also like it if decaffinate didn't add semicolons at the end of all my lines (my poor hipster JS style conventions...). Once you sort that out I'll merge the change.

A better way to verify immutability would be to recursively call `Object.freeze` on all inputs to all the functions in all tests. The nice thing about that approach is that if any tests inappropriately modify data, the stack trace will point to the exact source of the error. If we do that, we probably shouldn't need special immutability test cases at all.

curran commented:
Awesome! Thanks for reviewing. I will work on the points you mentioned.

I can tell Prettier not to include semicolons, so that change at least will be automatic.

Changing the way that the immutability tests work at first glance feels out of scope for this PR, but maybe it's necessary now if the decaffeination changed something subtle/fundamental with those particular tests.

josephg commented:
Yes I agree it’s out of scope. And I’d like to know what is causing the problem anyway, on the off chance it’s independently something worth fixing.

curran commented:
For the record, this is the prettier command I used:

`prettier --single-quote --no-semi ./*.js --write`

curran commented:
Ok I found the cause of the error.

The test uses a `require` call to pull in a `.coffee` file. In `test/genOp.js`:

```
194         const genString = require('ot-text/test/genOp') // Tries loading genOp.coffee
```

This should work if we bring back the `coffeescript/register` in mocha config.

It would not be a total decaffeination, but it's pretty close.

curran commented:
Woohoo CI passes! Onto the manual cleanup.

curran commented:
@josephg Ready for final review. I did a manual pass through the code and removed unnecessary stuff added by `decaffeinate` (mostly unnecessary `return` statements and IIFEs).

I would like to bring your attention to this line in the original test, which doesn't seem to make sense https://github.com/ottypes/json1/blob/d8c998d3ce9673591b155d6ea4e57dfb780e19db/test/test.coffee#L2003 .

```
      expect: [ [ 'a', { r: true }, 0, { p: 0 } ], [ 'b', { d: 0 } ] ]
```

I've added a `TODO` flag for you on the code generated from that line, right here https://github.com/ottypes/json1/pull/8/files#diff-c1129c8b045390789fa8ff62f2c6b4a9R2781

```
      // TODO have a look at this - doesn't seem right to just define an object
      {
        expect: [['a', { r: true }, 0, { p: 0 }], ['b', { d: 0 }]]
      }
```

Thanks for accepting my contribution!

curran commented on test/genOp.js:
```diff
@@ -0,0 +1,278 @@
+}
+
+const set = (container, key, value) =>
+  value === undefined
```
By the way, I manually changed a coupld of these `if` `else` sections to use these ternary operators instead. Happy to switch them back if you'd prefer.

josephg commented:
> I would like to bring your attention to this line in the original test, which doesn't seem to make sense
> 
> [json1/test/test.coffee](https://github.com/ottypes/json1/blob/d8c998d3ce9673591b155d6ea4e57dfb780e19db/test/test.coffee#L2003)
> Line 2003 in [d8c998d](/ottypes/json1/commit/d8c998d3ce9673591b155d6ea4e57dfb780e19db)
>  expect: [ [ 'a', { r: true }, 0, { p: 0 } ], [ 'b', { d: 0 } ] ] 
> .
> ```
>       expect: [ [ 'a', { r: true }, 0, { p: 0 } ], [ 'b', { d: 0 } ] ]
> ```

Yeah I have no idea what thats doing there. Its a remnant of some test case, no doubt. Its safe to delete.

I'm not entirely happy that coffeescript is still a dependancy after this change - though I assume we'll be able to remove it with a bit more refactoring.

But yeah, great stuff! I'm going to merge this.

curran commented:
Hooray!
